pallets/sgx-account-linker/src/lib.rs

```diff
@@ -43,172 +43,366 @@ type Signature = [u8; 65];
+            Self::do_link_eth(
+                account,
+                index,
+                addr_expected,
+                expiring_block_number,
+                layer_one_block_number,
+                sig,
+            )
+        }
+
+        /// Accept a pending `link_polkadot` request to link a Litentry address (= any account in Polkadot ecosystem)
+        /// to another Litentry address (= any account in Polkadot ecosystem).
+        ///
+        /// The runtime needs to ensure that a malicious index can be handled correctly.
+        /// Currently, when vec.len > MAX_ETH_LINKS, replacement will always happen at the final index.
+        /// Otherwise it will use the next new slot unless index is valid against a currently available slot.
+        ///
+        /// Parameters:
+        /// - `account`: The Litentry address that is to be linked
+        ///
+        /// Emits `PolkadotAddressLinked` event when successful.
+        // TODO will update weight when do the benchmark testing
+        #[pallet::weight(T::WeightInfo::link_eth())]
+        pub fn link_sub(
+            origin: OriginFor<T>,
+            account: T::AccountId,
+            index: u32,
+            network_type: PolkaNetType,
+            parachain_id: u32,
+            linked_account: T::AccountId,
+            layer_one_block_number: T::BlockNumber,
+            expiring_block_number: T::BlockNumber,
+            sig: MultiSignature,
+        ) -> DispatchResultWithPostInfo {
+            // in sgx runtime, the account who want to link ethereum address don't have the balance to
+            // submit extrinsic, the origin could be the root account
+            let _ = ensure_signed(origin)?;
+            Self::do_link_sub(
+                account,
+                index,
+                network_type,
+                parachain_id,
+                linked_account,
+                expiring_block_number,
+                layer_one_block_number,
+                sig,
+            )
+        }
+    }
+
+    impl<T: Config> Pallet<T> {
+        /// Assemble the message that the user has signed
+        /// Format: "Link Litentry: " + Litentry account + expiring block number
+        fn generate_raw_message(
+            account: &T::AccountId,
+            expiring_block_number: T::BlockNumber,
+        ) -> Vec<u8> {
+            let mut bytes = b"Link Litentry: ".encode();
+            let mut account_vec = account.encode();
+            let mut expiring_block_number_vec = expiring_block_number.encode();
+
+            bytes.append(&mut account_vec);
+            bytes.append(&mut expiring_block_number_vec);
+            bytes
+        }
+
+        /// Assemble the message that the user has signed
+        /// Format: "Link Litentry: " + network_type + parachain_id + Litentry account + expiring block number
+        fn generate_sub_raw_message(
+            account: &T::AccountId,
+            network_type: PolkaNetType,
+            parachain_id: u32,
```

Comment:
in my opinion parachain id here is unnecessary, as substrate wallet distinguishes networks only with network prefix. As we already list our supported network in PolkaNetType, which I think is a smart design, and the supported ones could be either parachain, or relay chain (polkadot, kusama) or even standalone substrate chain.

Reply:
will define as 
pub enum PolkaNetType {
        Kusama,
        Polkadot,
        KusamaParachain(i32),
        PolkadotParachain(i32),
    }